libmogwai-schedule/tests/signal-logger.h

```diff
@@ -0,0 +1,140 @@
+                             mws_signal_logger_get_n_emissions (self), \
+                             ane_list); \
+        g_assertion_message (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+                             ane_message); \
+    }\
+  } G_STMT_END
+
+/**
+ * mws_signal_logger_assert_emission_pop:
+ * @self: a #MwsSignalLogger
+ * @obj: a #GObject instance to assert the emission matches
+ * @signal_name: signal name to assert the emission matches
+ * @...: return locations for the signal parameters
+ *
+ * Assert that a signal emission can be popped off the log (using
+ * mws_signal_logger_pop_emission()) and that it is an emission of @signal_name
+ * on @obj. The parameters from the emission will be returned in the return
+ * locations given in the varargs, as with mws_signal_logger_get_params().
+ *
+ * If a signal emission can’t be popped, or if it doesn’t match @signal_name and
+ * @obj, an assertion fails, and some debug output is printed.
+ *
+ * Since: 0.1.0
+ */
+#define mws_signal_logger_assert_emission_pop(self, obj, signal_name, ...) \
+  G_STMT_START { \
+    gpointer aep_obj = NULL; \
+    g_autofree gchar *aep_obj_type_name = NULL; \
+    g_autofree gchar *aep_signal_name = NULL; \
+    g_autoptr(MwsSignalLoggerEmission) aep_emission = NULL; \
+    if (mws_signal_logger_pop_emission (self, &aep_obj, &aep_obj_type_name, \
+                                        &aep_signal_name, \
+                                        &aep_emission)) \
+      { \
+        if (aep_obj == G_OBJECT (obj) && \
+            g_str_equal (aep_signal_name, signal_name)) \
+          { \
+            /* Passed the test! */ \
+            mws_signal_logger_emission_get_params (aep_emission, __VA_ARGS__); \
+          } \
+        else \
+          { \
+            g_autofree gchar *aep_args = \
+                mws_signal_logger_format_emission (aep_obj,\
+                                                   aep_obj_type_name,\
+                                                   aep_signal_name,\
+                                                   aep_emission); \
+            g_autofree gchar *aep_message = \
+                g_strdup_printf ("Expected emission of %s::%s from %p, but saw: %s", \
+                                 G_OBJECT_TYPE_NAME (obj), signal_name, obj, \
+                                 aep_args); \
+            g_assertion_message (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+                                 aep_message); \
+          } \
+      } \
+    else \
+      { \
+        g_autofree gchar *assert_emission_pop_message = \
+            g_strdup_printf ("Expected emission of %s::%s from %p, but saw no emissions", \
+                             G_OBJECT_TYPE_NAME (obj), signal_name, obj); \
+        g_assertion_message (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+                             assert_emission_pop_message); \
+      } \
```

Comment:
Same comment about whether it's possible to split split part of this out into an actual function and calling it in the macro together with the message assertion parts.

Reply:
Same reply as above, although it does also annoy me that the macro is so long. I’ve taken another look at it and can’t figure out any of it which it would make sense to move into another public API. Suggestions welcome.

Reply:
I understand your point about not having a public function just as a helper. This is a tests header so I'd be fine with it (e.g. with a `_` prefix or so), but let's leave it like this then.